crates/primitives/src/bits/serde.rs

```diff
@@ -10,13 +14,83 @@ impl<const N: usize> Serialize for FixedBytes<N> {
 
 impl<'de, const N: usize> Deserialize<'de> for FixedBytes<N> {
     fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
-        hex::deserialize::<'de, D, [u8; N]>(deserializer).map(Self)
+        struct FixedVisitor<const N: usize>;
+
+        impl<'de, const N: usize> Visitor<'de> for FixedVisitor<N> {
+            type Value = FixedBytes<N>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(
+                    formatter,
+                    "{} bytes, represented as an array of u8, or a hex string of length {}",
+                    N,
+                    N * 2
+                )
+            }
+
+            fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                <[u8; N]>::try_from(v)
+                    .map(FixedBytes)
+                    .map_err(de::Error::custom)
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::SeqAccess<'de>,
+            {
+                let mut bytes = [0u8; N];
+
+                bytes.iter_mut().enumerate().try_for_each(|(i, b)| {
+                    *b = seq.next_element()?.ok_or_else(|| {
+                        de::Error::invalid_length(i, &format!("exactly {} bytes", N).as_str())
+                    })?;
+                    Ok(())
+                })?;
+
+                if let Ok(Some(_)) = seq.next_element::<u8>() {
+                    return Err(de::Error::invalid_length(
+                        N + 1,
+                        &format!("exactly {} bytes", N).as_str(),
+                    ))
+                }
+
+                Ok(FixedBytes(bytes))
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                let v = v.strip_prefix("0x").unwrap_or(v);
+
+                let v = hex::decode(v).map_err(|_| {
+                    de::Error::invalid_value(de::Unexpected::Str(v), &"a valid hex string")
+                })?;
+                <[u8; N]>::try_from(v.as_slice())
+                    .map_err(|_| {
+                        de::Error::invalid_length(
+                            v.len(),
+                            &format!("exactly {} bytes, as {} hex chars", N, N * 2).as_str(),
+                        )
+                    })
+                    .map(FixedBytes)
+            }
+        }
+        deserializer.deserialize_any(FixedVisitor::<N>)
```

Comment:
so we want to check for Human readable?

Reply:
i think that we should on Serialize. but on deserialize, because the only human-readable we expect is also self-describing (json), I think it's unnecessary. the serde_json deserializer will not treat a string as bytes unless explicitly instructed to via `deserialize_bytes`. That implies that checking for human-readable is unnecessary

Reply:
code ref: https://github.com/serde-rs/json/blob/master/src/de.rs#L1416-L1423